### PR TITLE
lib: lte_link_control: Add RAI notification support

### PR DESCRIPTION
**Pull request by trantanen**
Add support for new %RAI AT notifications.
Using AT%RAI=2 when enabling RAI to order those notifications. Fallback to AT%RAI=1 done for older modems.
These are mfw 2.0.2 features.

Support for these added also to modem shell.

Jira: NCSDK-26779

### CODE REVIEW

File: include/modem/lte_lc.h [hunk 1]
```diff
@@ -313,6 +313,16 @@ enum lte_lc_evt_type {
 	 * event.
 	 */
 	LTE_LC_EVT_MODEM_EVENT,
+
+	/**
+	 * Information about RAI (Release Assistance Indication) configuration.
+	 *
+	 * The associated payload is the @c lte_lc_evt.rai_cfg member of type
```
Comment (MirkoCovizzi) [inline include/modem/lte_lc.h:320]: Following the other docstrings:
```suggestion
	 * The associated payload is the @c rai_cfg member of type
```

File: lib/lte_link_control/lte_lc_helpers.c [hunk 2]
```diff
@@ -1543,3 +1594,29 @@ int parse_periodic_search_pattern(const char *const pattern_str,
...
+
+	err = nrf_modem_at_printf("AT%%RAI=%d", IS_ENABLED(CONFIG_LTE_RAI_REQ) ? 2 : 0);
+	if (err) {
+		if (IS_ENABLED(CONFIG_LTE_RAI_REQ)) {
+			LOG_DBG("Failed to enable RAI with notifications so trying without them");
+			/* If AT%RAI=2 failed, modem might not support it so using older API */
+			err = nrf_modem_at_printf("AT%%RAI=1");
+		}
+		if (err) {
+			LOG_ERR("Failed to configure RAI, err %d", err);
```
Comment (MirkoCovizzi) [inline lib/lte_link_control/lte_lc_helpers.c:1616]:
```suggestion
			LOG_ERR("Failed to enable RAI, err %d", err);
```

File: lib/lte_link_control/lte_lc_helpers.h [hunk 3]
```diff
@@ -307,6 +315,19 @@ int parse_coneval(const char *at_response, struct lte_lc_conn_eval_params *param
...
+/* @brief Parse a RAI response and extract RAI configuration.
+ *
+ * @param at_response AT response.
+ * @param rai_cfg RAI configuration.
+ *
+ * @return Zero on success, negative errno code on failure.
+ *
+ * @retval 0 Parsing succeeded.
+ * @retval -EINVAL If invalid parameters are provided.
+ * @retval -EIO If the AT response is not a valid RAI response.
```
Comment (MirkoCovizzi) [inline lib/lte_link_control/lte_lc_helpers.h:327]: Have you observed this error in practice? This error (from `at_parser`) indicates parsing beyond the limits of the AT response string.

File: lib/lte_link_control/lte_lc_helpers.c (hunk 2 above)
Comment (trantanen) [reply lib/lte_link_control/lte_lc_helpers.c:1616]: We do get here also when disabling is failing.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 4]
```diff
@@ -687,6 +687,10 @@ Modem libraries
       Use the :kconfig:option:`CONFIG_LTE_NETWORK_MODE_LTE_M_NBIOT` or :kconfig:option:`CONFIG_LTE_NETWORK_MODE_LTE_M_NBIOT_GPS` Kconfig option instead.
       In addition, you can control the priority between LTE-M and NB-IoT using the :kconfig:option:`CONFIG_LTE_MODE_PREFERENCE` Kconfig option.
 
+  * Added:
+
+    * A new :c:enum:`LTE_LC_EVT_RAI_UPDATE` event and subscribing to the event.
```
Comment (tokangas) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:692]: "subscribing to the event" is internal LTE LC functionality, needed to be able to send the new event. In my opinion it's unnecessary to mention it in the changelog.

File: lib/lte_link_control/lte_lc_helpers.c [hunk 5]
```diff
@@ -1429,6 +1430,56 @@ int parse_mdmev(const char *at_response, enum lte_lc_modem_evt *modem_evt)
...
+	__ASSERT_NO_MSG(err == 0);
+
+	/* Cell Id */
+	err = string_param_to_int(&parser, AT_RAI_CELL_ID_INDEX, &tmp_int, 16);
+	if (err) {
+		LOG_ERR("Could not parse cell_id, error: %d", err);
+		goto clean_exit;
+	}
+
+	if (tmp_int > LTE_LC_CELL_EUTRAN_ID_MAX) {
```
Comment (tokangas) [inline lib/lte_link_control/lte_lc_helpers.c:1449]: Is it expected that modem sends the notification with an invalid cell ID? If this is not expected behavior, do we want to send the event with an invalid cell ID, or should we rather just ignore the notification?

File: lib/lte_link_control/lte_lc_helpers.h (hunk 3 above)
Comment (tokangas) [reply lib/lte_link_control/lte_lc_helpers.h:327]: `at_parser` can return quite many different errors, not sure which ones are possible in this case. Do we need to try to document those here, after all, this is an internal API? I think at least `-EIO` could be removed.

File: lib/lte_link_control/lte_lc_helpers.c [hunk 6]
```diff
@@ -1543,3 +1594,29 @@ int parse_periodic_search_pattern(const char *const pattern_str,
 
 	return 0;
 }
+
+int lte_lc_rai_set(void)
```
Comment (tokangas) [inline lib/lte_link_control/lte_lc_helpers.c:1598]: This function is named like functions in the public API. Other functions in the helpers module do not have the `lte_lc` prefix, so I think it could be removed from this one as well to avoid confusion. I think `rai_set()` or `rai_configure()` would be enough.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 4 above)
Comment (trantanen) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:692]: I added so that users know that it's taken care of and it's not completely internal because user will select whether to subscribe with `CONFIG_LTE_RAI_REQ `. Maybe that existing option could be mentioned here too.

File: lib/lte_link_control/lte_lc_helpers.h (hunk 3 above)
Comment (trantanen) [reply lib/lte_link_control/lte_lc_helpers.h:327]: In unit tests I am able to get `-EIO` by giving RAI notification without some of the parameters so we try to read parameters beyond the limits. Obviously, this would be a modem bug if it would send such a notification.

Comment (trantanen) [reply lib/lte_link_control/lte_lc_helpers.h:327]: And generally, I this whole file has mixed styles. Some functions lists possible error codes while others don't. And I doubt there are some where the list is not complete.
I'm not planning to clean this up in this PR.

File: lib/lte_link_control/lte_lc_helpers.c (hunk 6 above)
Comment (trantanen) [reply lib/lte_link_control/lte_lc_helpers.c]: Renamed.

File: lib/lte_link_control/lte_lc_helpers.c (hunk 5 above)
Comment (trantanen) [reply lib/lte_link_control/lte_lc_helpers.c:1449]: I hope they don't send it with invalid CELL ID but I don't know. I can ignore it.

Comment (trantanen) [reply lib/lte_link_control/lte_lc_helpers.c:1449]: Ignored.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst (hunk 4 above)
Comment (trantanen) [reply doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:692]: Changed.

File: lib/lte_link_control/lte_lc_helpers.h (hunk 3 above)
Comment (trantanen) [reply lib/lte_link_control/lte_lc_helpers.h:327]: I've removed the comments related to return values as we are not sure if they are complete and they are not documented for all other functions either. This is an internal API anyway.

**Review by tokangas — APPROVED** [review]

File: include/modem/lte_lc.h [hunk 7]
```diff
@@ -209,40 +209,41 @@ enum lte_lc_evt_type {
 	/**
 	 * Network registration status.
 	 *
-	 * The associated payload is the @c nw_reg_status member of type @ref lte_lc_nw_reg_status
-	 * in the event.
+	 * The associated payload is the @c lte_lc_evt.nw_reg_status member of type
```
Comment (MirkoCovizzi) [inline include/modem/lte_lc.h:212]: I think that this and similar changes should be in a separate commit.

File: lib/lte_link_control/lte_lc_helpers.h [hunk 8]
```diff
@@ -372,3 +389,13 @@ char *periodic_search_pattern_get(char *const buf, size_t buf_size,
  */
 int parse_periodic_search_pattern(const char *const pattern_str,
 				  struct lte_lc_periodic_search_pattern *pattern);
+
+/* @brief Set RAI based on @kconfig{CONFIG_LTE_RAI_REQ}.
+ *
+ * If enabling of RAI is requested, AT%RAI=2 is used to order unsolicited RAI notifications.
```
Comment (MirkoCovizzi) [inline lib/lte_link_control/lte_lc_helpers.h:395]: I think the sentence structure can be improved here. For example:
```suggestion
 * When enabling RAI, it sets AT%RAI=2 which also enables RAI notifications. If that fails, falls back to AT%RAI=1.
```

**Review by MirkoCovizzi — COMMENTED** [review]
Looks good, just a couple of comments.

File: lib/lte_link_control/lte_lc_helpers.h (hunk 8 above)
Comment (MirkoCovizzi) [reply lib/lte_link_control/lte_lc_helpers.h:395]: Any suggestions @divipillai ?

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 9]
```diff
@@ -687,6 +687,10 @@ Modem libraries
       Use the :kconfig:option:`CONFIG_LTE_NETWORK_MODE_LTE_M_NBIOT` or :kconfig:option:`CONFIG_LTE_NETWORK_MODE_LTE_M_NBIOT_GPS` Kconfig option instead.
       In addition, you can control the priority between LTE-M and NB-IoT using the :kconfig:option:`CONFIG_LTE_MODE_PREFERENCE` Kconfig option.
 
+  * Added:
+
+    * A new :c:enum:`LTE_LC_EVT_RAI_UPDATE` event, which is enabled with the :kconfig:option:`CONFIG_LTE_RAI_REQ`.
```
Comment (peknis) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:692]:
```suggestion
    * A new :c:enum:`LTE_LC_EVT_RAI_UPDATE` event that is enabled with the :kconfig:option:`CONFIG_LTE_RAI_REQ` Kconfig option.
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: lib/lte_link_control/lte_lc_helpers.h (hunk 8 above)
Comment (trantanen) [reply lib/lte_link_control/lte_lc_helpers.h:395]: @peknis, please comment.

Comment (peknis) [reply lib/lte_link_control/lte_lc_helpers.h:395]: I'd be happy to approve the original wording.

**Review by MirkoCovizzi — APPROVED** [review]

**Review by peknis — APPROVED** [review]

File: include/modem/lte_lc.h [hunk 10]
```diff
@@ -209,40 +209,41 @@ enum lte_lc_evt_type {
...
 	LTE_LC_EVT_RRC_UPDATE,
 
 	/**
 	 * Current cell.
 	 *
-	 * The associated payload is the @c cell member of type @ref lte_lc_cell in the event.
-	 * Only the @c cell.tac and @c cell.id members are populated in the event payload. The rest
-	 * are expected to be zero.
+	 * The associated payload is the @c lte_lc_evt.cell member of type
+	 * @ref lte_lc_cell in the event. Only the @c lte_lc_cell.tac and @c lte_lc_cell.id
```
Comment (trantanen) [inline include/modem/lte_lc.h:245]: With latest updates, I've just added proper references for tac, cell ID and some time fields further below. Other than that the PR hasn't been touched. Now I'm waiting for the CI and doc build to confirm that references look good in generated doc.

**Comment by trantanen** [comment]
> We were thinking that it would be a good idea to add a simple RAI notification handling in the `udp` sample since we already have something for RAI there.

How do you want it to be used? Print RAI support information from the notification?

**Review by nordicjm — APPROVED** [review]
Build system changes OK

**Comment by MirkoCovizzi** [comment]
> > We were thinking that it would be a good idea to add a simple RAI notification handling in the `udp` sample since we already have something for RAI there.
> 
> How do you want it to be used? Print RAI support information from the notification?

Yes, something simple like that would be interesting to have in the sample.

**Review by MirkoCovizzi — APPROVED** [review]
